econml/inference.py

```diff
@@ -24,12 +32,59 @@ def __init__(self, n_bootstrap_samples=100, n_jobs=-1):
         self._n_bootstrap_samples = n_bootstrap_samples
         self._n_jobs = n_jobs
 
-    @property
-    def n_bootstrap_samples(self):
-        """Get how many draws to perform."""
-        return self._n_bootstrap_samples
+    def fit(self, estimator, *args, **kwargs):
+        est = BootstrapEstimator(estimator, self._n_bootstrap_samples, self._n_jobs, compute_means=False)
+        est.fit(*args, **kwargs)
+        self._est = est
+
+    def __getattr__(self, name):
+        m = getattr(self._est, name)
+
+        def wrapped(*args, alpha=0.1, **kwargs):
+            return m(*args, lower=100 * alpha / 2, upper=100 * (1 - alpha / 2), **kwargs)
+        return wrapped
+
+
+class StatsModelsInference(Inference):
+    """
+    Stores statsmodels covariance options.
+
+    This class can be used for inference by the LinearDMLCateEstimator.
+
+    Parameters
+    ----------
+    cov_type : string, optional (default 'HC1')
+        The type of covariance estimation method to use.  Supported values are 'nonrobust',
+        'fixed scale', 'HC0', 'HC1', 'HC2', and 'HC3'.  See the statsmodels documentation for
+        further information.
+
+    cov_kwds : optional additional keywords supported by the chosen method
+        Of the supported types, only 'fixed scale' has any keywords:
+        the optional keyword 'scale' which defaults to 1 if not specified.  See the statsmodels
+        documentation for further information.
+    """
+
+    def __init__(self, cov_type='HC1', **cov_kwds):
+        if cov_kwds and cov_type != 'fixed scale':
+            raise ValueError("Keyword arguments are only supported by the 'fixed scale' cov_type")
+        if cov_type not in ['nonrobust', 'fixed scale', 'HC0', 'HC1', 'HC2', 'HC3']:
+            raise ValueError("Unsupported cov_type; "
+                             "must be one of 'nonrobust', "
+                             "'fixed scale', 'HC0', 'HC1', 'HC2', or 'HC3'")
+        self.cov_type = cov_type
+        self.cov_kwds = cov_kwds
+
+    def fit(self, estimator, *args, **kwargs):
+        estimator._model_final.fit_args = {'cov_type': self.cov_type, 'cov_kwds': self.cov_kwds}
+        self._est = estimator
+
+    def effect_interval(self, X, *, T0=0, T1=1, alpha=0.1):
+        dT = T1 - T0
+        preds = self._est._model_final._model.predict_interval(self._est._model_final._combine(X, dT), alpha=alpha)
+        return preds
 
-    @property
-    def n_jobs(self):
-        """Get the maximum number of concurrently running jobs, as in joblib.Parallel."""
-        return self._n_jobs
+    def const_marginal_effect_interval(self, X, *, alpha=0.1):
+        # TODO: is there a cleaner way to achieve this?
+        XT = self._est._model_final._transform(X)
+        preds = self._est._model_final._model.predict_interval(XT, alpha=alpha)
```

Comment:
One thing I don't like here is that we are using dml specific properties. I think a statsmodel inference should be a general inference method that can work with any cate_estimator whose final effect model is a statsmodel linear regression. 

One approach: the statsmodelinference class takes as an extra init input a pointer to the final_model object of the cate estimator object that it wraps. 

Second approach (might be preferable): we uniformize across all of our estimators a way to access the final model, i.e. a model whose "predict(X)" yields theta(X). This requires that we first make sure that this is applicable to all our classes. Or at least whenever it is applicable we expose it in the same way.

Second the above would require that the _tranform(X) and _combine(X, dT) logic be moved somewhere else other than through inside DML. 

It feels that maybe we can uniformize all these by moving all this logic to linearCATE Estimator and then have the contract that StatsModelLinearRegression can only apply to linear CATE estimators. Then linear cate estimators should have a default way of accessing the final model. This could potentially be altered by child classes if needed, but it should be the same interface (e.g. sth like an effect_model method). 

This uniform way of accessing the final model would also be very useful for exposing the model for interpretability. I did implement such a method in the dmliv prototype, but can't say I did it in the most uniform way.

Reply:
I've been looking at this.  Basically, here's the problem (as I see it): for confidence intervals, the intervals for `effect(X, T0, T1)` and `const_marginal_effect(X)` are not cleanly related (particularly in the multiple treatment case).  We need at least the following ingredients:

* a way to get the statsmodels model that we want to use
* A way to combine X and T to get the actual inputs to the model
  * This can then be used directly with the model for computing `effect_interval`
* A way to produce the set of all basis unit treatments
  * Then we can replicate this set of basis treatments per row of X and repeat the rows of X and then combine these new arrays using the previous component and run the model to compute `const_marginal_effect`/`const_marginal_effect_interval` (and then if there are also multiple outputs, transpose the resulting array to get the shape/order right)

While in some sense these are the minimal components needed, in practice the set of steps in the last bullet (repeating things, combining them, transposing the result, etc.) are all identical for both `const_marginal_effect` (which is built on `predict` in `DMLCateEstimator` and `const_marginal_effect_interval` (which lives in `StatsModelsInference`).  Since it seems silly to reimplement this logic in both places, we're currently defining it in `DMLCateEstimator` and reaching into that logic from `StatsModelInference`.

Logically, it might be slightly nicer to have the initializer for `StatsModelInference` take these methods as parameters so that it never directly uses the internals of `DMLCateEstimator`, which should be a straightforward change.  However, it still feels like `const_marginal_effect_interval` is going to be surprisingly complex considering how conceptually simple it should be.

Reply:
@vasilismsr I've tried to address this over the last few commits.  I'm not completely sure this is the best possible organization, but I think it's sufficient for now and hopefully any future changes would only affect internals without affecting the interface exposed to users.  Let me know what you think.

Reply:
This has subsequently been further revised and now seems pretty reasonable.